### PR TITLE
[rllib] Also refactor DQN to use shared RLlib models

### PR DESCRIPTION
**Pull request by ericl**
This refactors the DQN implementation to use shared RLlib models. As part of this, I basically had to remove the dependence of tfutils which is good since it removes a bunch of code too.

I've done a cursory audit of the before and after tensor graphs and the changes look correct. More detailed performance evaluation:

## CartPole-v0:
CartPole seems fundamentally unstable with DQN (the baseline implementation exits as soon as you reach 200, but often if you continue learning the score collapses). Here I compared master vs. this refactor: there doesn't seem to be any significant difference in learning:

pole

## Pong-v0:
The learning curve is slightly different, but it does eventually reach +20 score in steady state.
pong

### CODE REVIEW

File: python/ray/rllib/dqn/example-cartpole.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+from __future__ import print_function
+
+import tensorflow as tf
+
+from ray.rllib.dqn import DQN, DEFAULT_CONFIG
+
+
+def main():
+    config = DEFAULT_CONFIG.copy()
+    config.update(dict(
```
Comment (ericl) [inline python/ray/rllib/dqn/example-cartpole.py]: Note: we should make train.py importable so that it can be called from these example scripts.

**Comment by robertnishihara** [comment]
Looks like there is a Python 2 error in the jenkins test. It wasn't introduced here, but it's probably worth fixing in this PR, what do you think @ericl?

```
Traceback (most recent call last):
  File "/ray/python/ray/rllib/dqn/example-cartpole.py", line 10, in <module>
    from ray.rllib.dqn import DQN, DEFAULT_CONFIG
  File "/opt/conda/lib/python2.7/site-packages/ray-0.1.2-py2.7-linux-x86_64.egg/ray/rllib/dqn/__init__.py", line 5, in <module>
    from ray.rllib.dqn.dqn import DQN, DEFAULT_CONFIG
  File "/opt/conda/lib/python2.7/site-packages/ray-0.1.2-py2.7-linux-x86_64.egg/ray/rllib/dqn/dqn.py", line 12, in <module>
    from ray.rllib.dqn import logger, models
  File "/opt/conda/lib/python2.7/site-packages/ray-0.1.2-py2.7-linux-x86_64.egg/ray/rllib/dqn/logger.py", line 143
    def log(*args, level=INFO):
                       ^
SyntaxError: invalid syntax
```

**Comment by pcmoritz** [comment]
Looks good to me as far as I can see; let's fix Python 2 and get it merged.

I'm really glad about getting rid of the tfutils, very nice :)

**Comment by pcmoritz** [comment]
retest this please

**Comment by pcmoritz** [comment]
Hm, looks like this should be fixed yet jenkins is complaining. Does anybody have an idea what is going on? @robertnishihara @shaneknapp

**Comment by robertnishihara** [comment]
There are more python 2 issues.

**Comment by ericl** [comment]
Hm my python2 installation is borked at the moment, but hopefully this latest push was the last issue.

retest this please

**Comment by robertnishihara** [comment]
The error this time is

```
Traceback (most recent call last):
  File "/ray/python/ray/rllib/dqn/example-cartpole.py", line 43, in <module>
    main()
  File "/ray/python/ray/rllib/dqn/example-cartpole.py", line 33, in main
    dqn = DQN("CartPole-v0", config)
  File "/opt/conda/lib/python2.7/site-packages/ray-0.1.2-py2.7-linux-x86_64.egg/ray/rllib/dqn/dqn.py", line 106, in __init__
    Algorithm.__init__(self, env_name, config, upload_dir=upload_dir)
  File "/opt/conda/lib/python2.7/site-packages/ray-0.1.2-py2.7-linux-x86_64.egg/ray/rllib/common.py", line 93, in __init__
    prefix=prefix, dir="/tmp/ray")
  File "/opt/conda/lib/python2.7/tempfile.py", line 339, in mkdtemp
    _os.mkdir(file, 0700)
OSError: [Errno 2] No such file or directory: '/tmp/ray/CartPole-v0_DQN_2017-07-26_15-25-33E95SEv'
```

It looks like the problem is that the directory `/tmp/ray` is created in `ray.init`, but I think we are attempting to create a file in that directory in the `DQN` constructor, which happens before the call to `ray.init`. In fact, currently there is no call to `ray.init`.

We could add a call to `ray.init` before creating the DQN object for now, but we'll still run into this problem when they happen out of order. Maybe we should just additionally attempt to create the `/tmp/ray` directory in the `Algorithm` constructor when we use it.

**Pull request merged**